### PR TITLE
[x-update-baseline] Initial implementation

### PR DESCRIPTION
**Pull request by strega-nil**
This is an initial, example implementation of `vcpkg x-update-baseline`; it is definitely incomplete, may be buggy, and could really use some progress messages

### CODE REVIEW

File: include/vcpkg/base/json.h [hunk 1]
```diff
@@ -213,6 +213,7 @@ namespace vcpkg::Json
 
         // replaces the value if the key is found, otherwise inserts a new
         // value.
+        Value& insert_or_replace(std::string key, std::string value);
```
Comment (BillyONeal) [inline include/vcpkg/base/json.h:216]: I thought we intentionally wanted callers to need to say Json::Value::whatever to make sure there was no confusion here?

File: src/vcpkg/base/json.cpp [hunk 2]
```diff
@@ -366,6 +366,20 @@ namespace vcpkg::Json
...
+        if (v)
+        {
+            *v = Json::Value::string(std::move(value));
+            return *v;
+        }
+        else
+        {
+            underlying_.push_back({std::move(key), Json::Value::string(std::move(value))});
+            return underlying_.back().second;
+        }
```
Comment (BillyONeal) [inline src/vcpkg/base/json.cpp:381]:
```suggestion
        return insert_or_replace(std::move(key), Json::Value::string(std::move(value)));
```

This does the "two-string-moves-is-often-worse-than-copy" antipattern. But as long as it's here we may as well avoid duplicate code.

**Review by BillyONeal — COMMENTED** [review]
(Not done reviewing yet.)

File: include/vcpkg/base/json.h (hunk 1 above)
Comment (ras0219-msft) [reply include/vcpkg/base/json.h:216]: At least in my opinion, strings are not confusing. They're also _so_ common that adding this overload significantly simplifies the rest of the codebase.

Comment (BillyONeal) [reply include/vcpkg/base/json.h:216]: I think it's fine too, but since it looked like this had been a deliberate design choice before I just wanted to highlight it as a change in policy.

Comment (strega-nil-ms) [reply include/vcpkg/base/json.h:216]: it was not a deliberate design choice - the reason I did this was due to overload resolution BS, iirc.

File: src/vcpkg/commands.update-baseline.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                Checks::check_exit(VCPKG_LINE_INFO, configuration_file.empty());
+                config_in_manifest = true;
+            }
+        }
+
+        if (configuration_file.empty() && !has_builtin_baseline && !config_in_manifest)
+        {
+            msg::println(Color::warning, msgUpdateBaselineNoConfiguration);
+            Checks::exit_success(VCPKG_LINE_INFO);
+        }
```
Comment (vicroms) [inline src/vcpkg/commands.update-baseline.cpp:101]: This prevents from inserting a built-in baseline if there's none in the manifest. Is that intended?

File: src/vcpkg/commands.update-baseline.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            {
+                reg.baseline = std::move(*p);
+            }
+        }
+
+        if (config_in_manifest)
+        {
+            manifest.insert_or_replace("vcpkg-configuration", configuration.serialize());
+        }
+        else
```
Comment (vicroms) [inline src/vcpkg/commands.update-baseline.cpp:133]:
```suggestion
        else if (!configuration_file.empty())
```

In general, I think all these conditionals need a re-work. I suggest checking whether a configuration file exists first and then try the fallbacks (embedded config, `builtin-baseline`).

**Review by vicroms — CHANGES_REQUESTED** [review]

File: src/vcpkg/commands.update-baseline.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        {
+            msg::println(Color::warning, msgUpdateBaselineGitError, msg::url = url);
+            msg::write_unlocalized_text_to_stdout(Color::warning, res.error());
+            return nullopt;
+        }
+    }
+
+    static Optional<std::string> get_latest_baseline(const VcpkgPaths& paths, const RegistryConfig& r)
```
Comment (ras0219-msft) [inline src/vcpkg/commands.update-baseline.cpp:44]: I think this code should live inside `configuration.cpp` to minimize the number of files that must be aware of different flavors of registries. It should offer something like

```cpp
/// <returns>nullopt on success</returns>
Optional<LocalizedString> update_baselines(const VcpkgPaths& paths, RegistryConfig& config_to_update);
```

To update builtin_baseline, this will have to synthesize a temporary config object, but I think that's pretty simple and clean.

File: src/vcpkg/commands.update-baseline.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    DECLARE_AND_REGISTER_MESSAGE(UpdateBaselineNoConfiguration, (), "", "There is no configuration file to update.");
+    DECLARE_AND_REGISTER_MESSAGE(UpdateBaselineGitError,
+                                 (msg::url),
+                                 "",
+                                 "Git failed to fetch remote repository {url}:");
+}
+
+namespace vcpkg::Commands
+{
+    static const CommandStructure COMMAND_STRUCTURE{
```
Comment (ras0219-msft) [inline src/vcpkg/commands.update-baseline.cpp:21]: A dry run switch might be useful.

**Comment by ras0219-msft** [comment]
Moving to draft since there has been no update in >3wks

**Comment by strega-nil-ms** [comment]
I believe that I've fixed the CRs for both Victor and Robert; can you rereview?

**Comment by strega-nil-ms** [comment]
I'd also appreciate bikeshedding on the messages

File: include/vcpkg/base/checks.h [hunk 7]
```diff
@@ -34,6 +34,11 @@ namespace vcpkg::Checks
         msg_exit_with_message(line_info, msg::format(msg::msgErrorMessage).append(msg::format(m, args...)));
     }
 
+    [[noreturn]] inline void msg_exit_with_error(const LineInfo& line_info, const LocalizedString& s)
```
Comment (BillyONeal) [inline include/vcpkg/base/checks.h:37]: I think this is a merge conflict; I don't see this one called, everything I see here is calling the one on line 30.

File: src/vcpkg/commands.update-baseline.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+#include <vcpkg/vcpkgcmdarguments.h>
+#include <vcpkg/vcpkgpaths.h>
+
+namespace
+{
+    namespace msg = vcpkg::msg;
+    DECLARE_AND_REGISTER_MESSAGE(UpdateBaselineNoConfiguration,
+                                 (),
+                                 "",
+                                 "there is no `vcpkg-configuration.json` nor manifest file to update.");
```
Comment (BillyONeal) [inline src/vcpkg/commands.update-baseline.cpp:15]:
```suggestion
                                 "neither `vcpkg.json` nor `vcpkg-configuration.json` exist to update.");
```

I think we've been referring to "the manifest" as "vcpkg.json + vcpkg-configuration.json" ? (And also sometimes control files...)

I think just naming the thing will be clearer for folks.

File: src/vcpkg/commands.update-baseline.cpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+#include <vcpkg/base/messages.h>
+
+#include <vcpkg/commands.update-baseline.h>
+#include <vcpkg/configuration.h>
+#include <vcpkg/vcpkgcmdarguments.h>
+#include <vcpkg/vcpkgpaths.h>
+
+namespace
+{
+    namespace msg = vcpkg::msg;
```
Comment (BillyONeal) [inline src/vcpkg/commands.update-baseline.cpp:11]:
```suggestion
    using namespace vcpkg;
```

File: src/vcpkg/commands.update-baseline.cpp [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    void UpdateBaselineCommand::perform_and_exit(const VcpkgCmdArguments& args, const VcpkgPaths& paths) const
+    {
+        auto options = args.parse_arguments(COMMAND_STRUCTURE);
+
+        const bool add_builtin_baseline = Util::Sets::contains(options.switches, OPTION_ADD_INITIAL_BASELINE);
+        const bool dry_run = Util::Sets::contains(options.switches, OPTION_DRY_RUN);
+
+        auto configuration = paths.get_configuration();
+        const bool has_manifest = paths.get_manifest().has_value();
+        auto manifest = has_manifest ? *paths.get_manifest().get() : ManifestAndLocation{};
```
Comment (BillyONeal) [inline src/vcpkg/commands.update-baseline.cpp:93]: Would this be clearer as an Optional<ManifestAndLocation> ? (Genuinely asking if you think so, not asking to do it; I have no idea if that'd be better or not)

File: src/vcpkg/commands.update-baseline.cpp [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+            auto old_baseline = p_builtin_baseline ? Optional<StringView>(p_builtin_baseline->string()) : nullopt;
+            print_update_message(synthesized_registry.registry_location(), old_baseline, result);
+
+            if (auto p = result.get())
+            {
+                manifest.manifest.insert_or_replace("builtin-baseline", std::move(*p));
+            }
+            else if (!error.empty())
+            {
+                msg::print_warning(error);
```
Comment (BillyONeal) [inline src/vcpkg/commands.update-baseline.cpp:136]: Is it a warning or an error?

File: src/vcpkg/configuration.cpp [hunk 12]
```diff
@@ -535,6 +537,66 @@ namespace
 
 namespace vcpkg
 {
+    static Optional<std::string> get_baseline_from_git_repo(const VcpkgPaths& paths,
+                                                            StringView url,
+                                                            LocalizedString& error)
```
Comment (BillyONeal) [inline src/vcpkg/configuration.cpp:542]:
```suggestion
    static ExpectedL<std::string> get_baseline_from_git_repo(const VcpkgPaths& paths,
                                                            StringView url)
```

Comment (BillyONeal) [reply src/vcpkg/configuration.cpp:542]: ExpectedL has better behavior than Optional if you `value_or_exit` it, for example: printing the error text instead of "value was null"

File: src/vcpkg/configuration.cpp [hunk 13]
```diff
@@ -535,6 +537,66 @@ namespace
...
+            error = msg::format(msgUpdateBaselineGitError, msg::url = url)
+                        .appendnl()
+                        .append_raw(Strings::trim(res.error()));
+            return nullopt;
+        }
+    }
+
+    static constexpr StringLiteral BUILTIN_GIT_URL = "https://github.com/microsoft/vcpkg";
+
+    Optional<std::string> RegistryConfig::get_latest_baseline(const VcpkgPaths& paths, LocalizedString& error) const
```
Comment (BillyONeal) [inline src/vcpkg/configuration.cpp:560]: I'm not sure I understand the contract of this function. The others are one of two options:

1. The return value is engaged and the error is not set
2. The return value is disengaged and the error is set

which seem consistent. But this one has 

3. The return value is disengaged and the error is not set

and I don't know what to make of that.

(I guess what I'm looking for is a comment)

File: src/vcpkg/commands.update-baseline.cpp (hunk 10 above)
Comment (strega-nil-ms) [reply src/vcpkg/commands.update-baseline.cpp:93]: No, it would not (it was before and I changed it)

File: src/vcpkg/commands.update-baseline.cpp (hunk 11 above)
Comment (strega-nil-ms) [reply src/vcpkg/commands.update-baseline.cpp:136]: it is a warning, because an error for _this_ function ends up just being a warning.

**Review by BillyONeal — APPROVED** [review]
Wow I'm surprised you did the wrapped expected thing but sounds good to me

**Comment by ras0219-msft** [comment]
Still working on my review, please do not merge yet.

File: src/vcpkg/configuration.cpp [hunk 14]
```diff
@@ -535,6 +537,61 @@ namespace
...
+        }
+        else
+        {
+            return msg::format(msgUpdateBaselineGitError, msg::url = url)
+                .appendnl()
+                .append_raw(Strings::trim(res.error()));
+        }
+    }
+
+    static constexpr StringLiteral BUILTIN_GIT_URL = "https://github.com/microsoft/vcpkg";
```
Comment (ras0219-msft) [inline src/vcpkg/configuration.cpp:555]: This should be deduplicated against the string in src/vcpkg/registries.cpp:1249 (`vcpkg::make_builtin_registry()`)

File: src/vcpkg/configuration.cpp [hunk 15]
```diff
@@ -535,6 +537,61 @@ namespace
...
+
+    ExpectedL<Optional<std::string>> RegistryConfig::get_latest_baseline(const VcpkgPaths& paths) const
+    {
+        if (kind == RegistryConfigDeserializer::KIND_GIT)
+        {
+            return get_baseline_from_git_repo(paths, repo.value_or_exit(VCPKG_LINE_INFO));
+        }
+        else if (kind == RegistryConfigDeserializer::KIND_BUILTIN)
+        {
+            return get_baseline_from_git_repo(paths, BUILTIN_GIT_URL);
```
Comment (ras0219-msft) [inline src/vcpkg/configuration.cpp:565]: This branch should get `HEAD` from the actual repo that will be used. This could mean `https://github.com/Microsoft/vcpkg` (if `use_git_default_registry()`), but it most likely means the _local_ git repo.

File: src/vcpkg/configuration.cpp [hunk 16]
```diff
@@ -535,6 +537,61 @@ namespace
...
+        {
+            return baseline;
+        }
+    }
+
+    StringView RegistryConfig::registry_location() const
+    {
+        if (kind == RegistryConfigDeserializer::KIND_BUILTIN)
+        {
+            return BUILTIN_GIT_URL;
```
Comment (ras0219-msft) [inline src/vcpkg/configuration.cpp:577]: Since this is only used for "pretty viewing" by the user, I think this can stay as the upstream URL.

File: src/vcpkg/update.cpp [hunk 17]
```diff
@@ -54,13 +54,18 @@ namespace vcpkg::Update
...
+
     void perform_and_exit(const VcpkgCmdArguments& args, const VcpkgPaths& paths)
     {
         if (paths.manifest_mode_enabled())
         {
-            Checks::exit_maybe_upgrade(VCPKG_LINE_INFO,
-                                       "Error: the update command does not currently support manifest mode. Instead, "
-                                       "modify your vcpkg.json and run install.");
+            msg::println(Color::error, msgUpdateCommandNoManifestMode);
+            Checks::msg_exit_maybe_upgrade(VCPKG_LINE_INFO, msgUpdateCommandDidYouMean);
```
Comment (ras0219-msft) [inline src/vcpkg/update.cpp:68]: This would be a good separate PR since there's no reason to touch this file here.

File: src/vcpkg/update.cpp [hunk 18]
```diff
@@ -54,13 +54,18 @@ namespace vcpkg::Update
         nullptr,
     };
 
+    DECLARE_AND_REGISTER_MESSAGE(UpdateCommandNoManifestMode,
+                                 (),
+                                 "",
+                                 "Error: the update command does not support manifest mode.");
```
Comment (ras0219-msft) [inline src/vcpkg/update.cpp:60]: Shouldn't this use `msg::msgErrorMessage`?

File: include/vcpkg/configuration.h [hunk 19]
```diff
@@ -23,6 +23,9 @@ namespace vcpkg
         Optional<std::vector<std::string>> packages;
 
         Json::Value serialize() const;
+
+        Optional<std::string> get_latest_baseline(const VcpkgPaths& paths, LocalizedString& error) const;
+        StringView registry_location() const;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/configuration.h:28]: (* requesting changes *)

File: include/vcpkg/configuration.h [hunk 20]
```diff
@@ -42,6 +45,25 @@ namespace vcpkg
         static View<StringView> known_fields();
     };
 
+    enum class ConfigurationLocation
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/configuration.h:48]: (* requesting changes *)

File: include/vcpkg/vcpkgpaths.h [hunk 21]
```diff
@@ -58,6 +58,12 @@ namespace vcpkg
     struct PackageSpec;
     struct Triplet;
 
+    struct ManifestAndLocation
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/vcpkgpaths.h:61]: (* requesting changes *)

File: src/vcpkg/configuration.cpp (hunk 15 above)
Comment (ras0219-msft) [reply src/vcpkg/configuration.cpp:565]: (* requesting changes *)

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
I've marked the comments provoking "request changes" with `(* requesting changes *)` -- my other review comments can be considered nits.

File: src/vcpkg/commands.update-baseline.cpp [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+            configuration.config.default_reg = nullopt;
+
+            auto synthesized_registry = RegistryConfig{};
+            synthesized_registry.kind = "builtin";
+
+            auto result = synthesized_registry.get_latest_baseline(paths, error);
+
+            auto p_builtin_baseline = manifest.manifest.get("builtin-baseline");
+            auto old_baseline = p_builtin_baseline ? Optional<StringView>(p_builtin_baseline->string()) : nullopt;
+            print_update_message(synthesized_registry.registry_location(), old_baseline, result);
```
[the remark this answers is not in the document]
Comment (strega-nil-ms) [reply src/vcpkg/commands.update-baseline.cpp]: out of date comment

File: src/vcpkg/configuration.cpp (hunk 15 above)
Comment (strega-nil-ms) [reply src/vcpkg/configuration.cpp:565]: I want to talk about this - I don't understand what this means.

Comment (strega-nil-ms) [reply src/vcpkg/configuration.cpp:565]: My understanding is that having a `builtin-baseline` implies using a git builtin registry...

**Review by BillyONeal — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
@ras0219-msft I believe I did what you wanted?

**Review by ras0219-msft — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
Thanks for the approvals y'all!
